{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from Bio import SeqIO, Phylo, Seq\n",
    "from string import digits\n",
    "import pandas as pd\n",
    "from collections import defaultdict, Counter\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "matrix unscaled synonymous\n",
      "      A     C     G     T\n",
      "A   NaN   249  2342   662\n",
      "C   589   NaN    28  4079\n",
      "G  2301    10   NaN   168\n",
      "T   541  4138   116   NaN\n",
      "\n",
      " syn  fractions\n",
      "0.34887810367869454\n",
      "\n",
      " scaled by syn ratio\n",
      "             A             C            G             T\n",
      "A     0.000000    713.716331  6712.946371   1897.510887\n",
      "C  1688.268750      0.000000    80.257258  11691.762702\n",
      "G  6595.426815     28.663306     0.000000    481.543548\n",
      "T  1550.684879  11860.876210   332.494355      0.000000\n",
      "\n",
      " ratio of nt \n",
      "\n",
      "          A         T        G         C\n",
      "0  0.384226  0.277973  0.16134  0.176461\n",
      "              A             C             G             T\n",
      "A      0.000000   1857.545055  17471.367549   4938.533440\n",
      "C   9567.377562      0.000000    454.815911  66256.932216\n",
      "G  40878.949886    177.657322      0.000000   2984.643016\n",
      "T   5578.544397  42669.162136   1196.138910      0.000000\n",
      "\n",
      " normalized and scaled matrix\n",
      "sum of all: 194031.6674024871 \n",
      "\n",
      "          A         C         G         T\n",
      "A  0.000000  0.009573  0.090044  0.025452\n",
      "C  0.049308  0.000000  0.002344  0.341475\n",
      "G  0.210682  0.000916  0.000000  0.015382\n",
      "T  0.028751  0.219908  0.006165  0.000000\n"
     ]
    }
   ],
   "source": [
    "def CDS_finder(reference):\n",
    "    \"\"\"this function finds CDS location and not CDS location, and saves only those not located at the end of the function\"\"\"\n",
    "    cds_ = dict()\n",
    "    for feature in reference.features:\n",
    "        if feature.type == 'CDS': \n",
    "            cds_[feature.qualifiers['gene'][0]] = (list(feature.location))\n",
    "    return(cds_)\n",
    "\n",
    "ref_file = SeqIO.read(\"data/areference.gbk\", \"genbank\")\n",
    "\n",
    "def Synonymous_Mutations(node, dictionary_=None, new_=None):\n",
    "    \"\"\" Finds Synonymous mutations in CDS regions. Input:nested json\"\"\"\n",
    "    gene_cds = CDS_finder(ref_file)\n",
    "    if new_ is None: new_ = []\n",
    "    if dictionary_ is None: dictionary_ = dict()\n",
    "    if 'mutations' in node['branch_attrs']:\n",
    "        aa_mutations, new_, in_it = ([] for i in range(3))\n",
    "        if 'nuc' in node['branch_attrs']['mutations']:\n",
    "            for gene, loc in gene_cds.items():\n",
    "                if gene in node['branch_attrs']['mutations']:\n",
    "                    for mut in node['branch_attrs']['mutations'][gene]:\n",
    "                        aa_mutations.append(int(mut[1:-1])*3+loc[0]) # converting the amino acid location to nucleotide\n",
    "            for mut in node['branch_attrs']['mutations']['nuc']:\n",
    "                if '-' not in mut and '*' not in mut and 'N' not in mut and \"R\" not in mut and \"Y\" not in mut and \"M\" not in mut and \"D\" not in mut:\n",
    "                    if int(mut[1:-1]) not in aa_mutations and int(mut[1:-1])+2 not in aa_mutations and int(mut[1:-1])+1 not in aa_mutations:  #if the mutation is not in the same codon as a aa mutation\n",
    "                        new_.append(mut)\n",
    "                    else: in_it.append(mut[1:-1])\n",
    "    if 'name' in node:\n",
    "            dictionary_[node['name']] = new_\n",
    "    if 'children' in node:\n",
    "        for child in node['children']:\n",
    "           Synonymous_Mutations(child, dictionary_, new_=None)\n",
    "    return(dictionary_)\n",
    "\n",
    "\n",
    "ref_file = SeqIO.read(\"data/areference.gbk\", \"genbank\")\n",
    "with open (\"data/rsv_a_genome.json\") as file_: f = json.load(file_)  \n",
    "synonymous = Synonymous_Mutations(f['tree'])\n",
    "\n",
    "\n",
    "\n",
    "#Extracting all synonymous mutations from json as a list\n",
    "all_muts = []\n",
    "for branch, muts in synonymous.items():\n",
    "    if branch != []:\n",
    "        for mut in muts:\n",
    "            if mut[0] and mut[-1] in [\"A\", \"T\", \"C\", \"G\"]:\n",
    "                all_muts.append(f'{mut[0]}{mut[-1]}')\n",
    "all_muts_counter = Counter(all_muts)\n",
    "\n",
    "#making matrix with mutation rates\n",
    "df = pd.DataFrame(index=['A', 'C', 'G', 'T'], columns=['A', 'C', 'G', 'T'])      \n",
    "\n",
    "for mutation, nr in all_muts_counter.items():\n",
    "    df.at[mutation[0], mutation[-1]] = int(nr)\n",
    "print(\"matrix unscaled synonymous\")\n",
    "print(df)\n",
    "\n",
    "\n",
    "#now have to divide by the nr of locations where the mutation can occur\n",
    "ref_file = SeqIO.read(\"data/areference.gbk\", \"genbank\")\n",
    "gene_cds = CDS_finder(ref_file)\n",
    "\n",
    "sequence_ref_cds = dict()\n",
    "whole_seq_CDS = \"\"\n",
    "for gene, cds in gene_cds.items(): \n",
    "    sequence_ref_cds[gene] = ref_file.seq[cds[0]:cds[-1]]\n",
    "    whole_seq_CDS = whole_seq_CDS+ref_file.seq[cds[0]:cds[-1]]\n",
    "whole_seq_CDS = whole_seq_CDS.replace(\"N\", \"\")\n",
    "whole_seq_CDS = whole_seq_CDS.replace(\"M\", \"\")\n",
    "\n",
    "#translation matrix\n",
    "translations = {'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'], 'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'], 'C': ['TGT', 'TGC'], 'W': ['TGG'], 'E': ['GAA', 'GAG'], 'D': ['GAT', 'GAC'], 'P': ['CCT', 'CCC', 'CCA', 'CCG'], 'V': ['GTT', 'GTC', 'GTA', 'GTG'], 'N': ['AAT', 'AAC'], 'M': ['ATG'], 'K': ['AAA', 'AAG'], 'Y': ['TAT', 'TAC'], 'I': ['ATT', 'ATC', 'ATA'], 'Q': ['CAA', 'CAG'], 'F': ['TTT', 'TTC'], 'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'], 'T': ['ACT', 'ACC', 'ACA', 'ACG'], '*': ['TAA', 'TAG', 'TGA'], 'A': ['GCT', 'GCC', 'GCA', 'GCG'], 'G': ['GGT', 'GGC', 'GGA', 'GGG'], 'H': ['CAT', 'CAC']}\n",
    "\n",
    "\n",
    "synonymous_possibilities, nonsynonymous_possibilities = (0 for i in range(2))\n",
    "for gene, sequence in sequence_ref_cds.items():\n",
    "    for i, letter in enumerate(sequence):\n",
    "        if i%3 == 0:\n",
    "            codon = sequence[i: i+3]\n",
    "            for key, entry in translations.items():\n",
    "                if codon in entry:\n",
    "                    synonymous = len(entry)\n",
    "                    synonymous_possibilities += synonymous-1 \n",
    "                    nonsynonymous = 9 - len(entry)-1\n",
    "                    nonsynonymous_possibilities += nonsynonymous\n",
    "\n",
    "nonsyn_ratio = nonsynonymous_possibilities/(nonsynonymous_possibilities+synonymous_possibilities)\n",
    "syn_ratio = synonymous_possibilities/(nonsynonymous_possibilities+synonymous_possibilities)\n",
    "\n",
    "print(\"\\n\",\"syn  fractions\")\n",
    "print(syn_ratio)\n",
    "\n",
    "print(\"\\n\",\"scaled by syn ratio\")\n",
    "scaled = df.divide(syn_ratio)\n",
    "scaled = scaled.fillna(0)\n",
    "print(scaled)\n",
    "\n",
    "print(\"\\n\", \"ratio of nt\", \"\\n\")\n",
    "counter = Counter(whole_seq_CDS)\n",
    "total = 0\n",
    "for i in counter.values(): total += i\n",
    "\n",
    "df_ratios = pd.DataFrame.from_dict(counter, orient='index').astype(int).T\n",
    "df_ratios = df_ratios.divide(total)\n",
    "print(df_ratios)\n",
    "\n",
    "nuc = [\"A\", \"C\", \"G\", \"T\"]\n",
    "for n in nuc:\n",
    "    scaled.loc[[n]] = scaled.loc[[n]].div(float(df_ratios[n]))\n",
    "\n",
    "print(scaled)\n",
    "\n",
    "print(\"\\n\", \"normalized and scaled matrix\")\n",
    "\n",
    "sum_ = scaled.to_numpy().sum()\n",
    "print(\"sum of all:\",  sum_, \"\\n\")\n",
    "\n",
    "scaled_and_normalized = scaled.divide(sum_)\n",
    "\n",
    "print(scaled_and_normalized)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextstrain",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
