{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from Bio import SeqIO, Phylo, Seq\n",
    "from string import digits\n",
    "from collections import defaultdict, Counter\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.1469072199999948\n",
      "nucleotide before T to C {'A': 0.8858364517246887, 'C': 1.7750104997364882, 'G': 0.8921508800984707, 'T': 0.7986962836899302}\n",
      "nucleotide before C to T {'T': 1.3640732062049785, 'C': 2.559275717665431, 'A': 1.4251365484155716, 'G': 1.1260229491513583}\n"
     ]
    }
   ],
   "source": [
    "def CDS_list(reference):\n",
    "    \"\"\"this function finds CDS location and not CDS location, and saves only those not located at the end of the function\"\"\"\n",
    "    cds_ = []\n",
    "    for feature in reference.features:\n",
    "        if feature.type == 'CDS': \n",
    "            cds_.extend(list(feature.location))\n",
    "    return(cds_)\n",
    "\n",
    "def before_and_after (reference, nr, location, nucl):\n",
    "        \"\"\"this function returns counts of nucleotides before or after a specific nucleotide in the CDS region of the reference\"\"\"\n",
    "\n",
    "        ref = SeqIO.read(reference, \"genbank\")\n",
    "        CDS_ = CDS_list(ref)\n",
    "        all_motifs = []\n",
    "        for i, nu in enumerate(ref.seq):\n",
    "            if nu == nucl and i in CDS_:    \n",
    "                if location == \"before\" and i-nr in CDS_:\n",
    "                    if \"N\" not in ref.seq[i-1]:\n",
    "                        all_motifs.append(ref.seq[i-nr])\n",
    "                if location == \"after\" and i+nr in CDS_:\n",
    "                    if 'N' not in ref.seq[i+1]:\n",
    "                     all_motifs.append(ref.seq[i+1])\n",
    "        return(Counter(all_motifs))\n",
    "\n",
    "\n",
    "def CDS_finder(reference):\n",
    "    \"\"\"this function finds CDS location and not CDS location, and saves only those not located at the end of the function\"\"\"\n",
    "    cds_ = dict()\n",
    "    for feature in reference.features:\n",
    "        if feature.type == 'CDS': \n",
    "            cds_[feature.qualifiers['gene'][0]] = (list(feature.location))\n",
    "    return(cds_)\n",
    "\n",
    "\n",
    "ref_file = SeqIO.read(\"data/areference.gbk\", \"genbank\")\n",
    "\n",
    "\n",
    "def mutation_recursive(node, dictionary_=None, new_=None):\n",
    "    \"\"\" this function returns a dictionary with node name as key and a list of mutations along that branch as the info, and excludes deletions and unknowns\"\"\"\n",
    "    gene_cds = CDS_finder(ref_file)\n",
    "    if new_ is None: new_ = []\n",
    "    if dictionary_ is None: dictionary_ = dict()\n",
    "    if 'mutations' in node['branch_attrs']:\n",
    "        aa_mutations, new_, in_it = ([] for i in range(3))\n",
    "        if 'nuc' in node['branch_attrs']['mutations']:\n",
    "            for gene, loc in gene_cds.items():\n",
    "                if gene in node['branch_attrs']['mutations']:\n",
    "                    for mut in node['branch_attrs']['mutations'][gene]:\n",
    "                        aa_mutations.append(int(mut[1:-1])*3+loc[0])\n",
    "            for mut in node['branch_attrs']['mutations']['nuc']:\n",
    "                if '-' not in mut and '*' not in mut and 'N' not in mut and \"R\" not in mut and \"Y\" not in mut and \"M\" not in mut and \"D\" not in mut:\n",
    "                    if int(mut[1:-1]) not in aa_mutations and int(mut[1:-1])+2 not in aa_mutations and int(mut[1:-1])+1 not in aa_mutations:\n",
    "                        new_.append(mut)\n",
    "                    else: in_it.append(mut[1:-1])\n",
    "    if 'name' in node:\n",
    "            dictionary_[node['name']] = new_\n",
    "    if 'children' in node:\n",
    "        for child in node['children']:\n",
    "           mutation_recursive(child, dictionary_, new_=None)\n",
    "    return(dictionary_)\n",
    "\n",
    "ref_file = SeqIO.read(\"data/areference.gbk\", \"genbank\")\n",
    "with open (\"data/rsv_a_genome.json\") as file_:\n",
    "    f = json.load(file_)  \n",
    "    mut_by_node = mutation_recursive(f['tree'])\n",
    "\n",
    "def find_relevant_nucl(sequencefile, dictionarysynonymous, motif, where):\n",
    "    \"\"\"this function takes as input reconstructed tree branches and a dictionary of synonymous mutations and returns nucleotides before or after the mutation of interest\"\"\"\n",
    "    records = SeqIO.parse(sequencefile, \"fasta\")\n",
    "    output = []\n",
    "    for record in records:\n",
    "        for mut in dictionarysynonymous[record.id]:\n",
    "            if f'{mut[0]}{mut[-1]}' == motif:\n",
    "                    location_of_mut = int(mut[1:-1])-1\n",
    "                    if where == \"after\":\n",
    "                        nucleotide = record.seq[location_of_mut+1]\n",
    "                        if nucleotide != '-' and nucleotide!= 'N': output.append(nucleotide)\n",
    "                    if where == \"before\":\n",
    "                        nucleotide = record.seq[location_of_mut-1]\n",
    "                        if nucleotide != '-' and nucleotide!= 'N': output.append(nucleotide)\n",
    "    return(Counter(output))\n",
    "\n",
    "\n",
    "all_= find_relevant_nucl(\"data/reconstructed_sequences.fasta\", mut_by_node, \"TC\", \"before\")\n",
    "scale = before_and_after(\"data/areference.gbk\", 1, \"before\", \"T\")\n",
    "# multiply by nr of branches\n",
    "\n",
    "sequences = SeqIO.parse(\"data/reconstructed_sequences.fasta\", \"fasta\")\n",
    "length = 0\n",
    "for i in sequences:\n",
    "    length+=1\n",
    "\n",
    "from Bio import Phylo\n",
    "\n",
    "tree_ = Phylo.read(\"data/a_tree.nwk\", \"newick\")\n",
    "treelength = tree_.total_branch_length()\n",
    "#here I divide the chances of it occurring by the number of branches in the tree\n",
    "#scaled_ ={x:float(all_[x])/(scale[x]*length) for x in all_}\n",
    "\n",
    "scaled_ ={x:float(all_[x])/(scale[x]*treelength) for x in all_}\n",
    "print(\"nucleotide before T to C\", scaled_)\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "all_= find_relevant_nucl(\"data/reconstructed_sequences.fasta\", mut_by_node, \"CT\", \"before\")\n",
    "scale = before_and_after(\"data/areference.gbk\", 1, \"before\", \"C\")\n",
    "scaled_ ={x:float(all_[x])/(scale[x]*treelength) for x in all_}\n",
    "print(\"nucleotide before C to T\", scaled_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         CT        CA        CG        TC        TA        TG        AC  \\\n",
      "T  1.364073  0.302975  0.019193  0.798696  0.177805  0.038033  0.035426   \n",
      "C  2.559276  0.401455  0.010455  1.775010  0.320409  0.052456  0.082893   \n",
      "A  1.425137  0.131470  0.003155  0.885836  0.050861  0.012715  0.040429   \n",
      "G  1.126023  0.107240  0.006994  0.892151  0.068507  0.026469  0.010061   \n",
      "\n",
      "         AG        AT        GC        GT        GA  \n",
      "T  0.489455  0.101940  0.004529  0.106441  0.777925  \n",
      "C  0.358435  0.290125  0.009797  0.215528  4.075445  \n",
      "A  0.435468  0.036533  0.001015  0.036541  0.920632  \n",
      "G  0.334232  0.070424  0.009634  0.038537  0.674406  \n"
     ]
    }
   ],
   "source": [
    "#dataframe of counts\n",
    "import pandas as pd\n",
    "combos = [\"CT\", \"CA\", \"CG\", \"TC\", \"TA\", \"TG\", \"AC\", \"AG\", \"AT\", \"GC\", \"GT\", \"GA\"]\n",
    "all_dictionaries = dict()\n",
    "for mut in combos:\n",
    "    all_= find_relevant_nucl(\"data/reconstructed_sequences.fasta\", mut_by_node, mut, \"before\")\n",
    "    scale = before_and_after(\"data/areference.gbk\", 1, \"before\", mut[0])\n",
    "    scaled_ ={x:float(all_[x])/(scale[x]*treelength) for x in all_}\n",
    "    all_dictionaries[mut]= scaled_\n",
    "\n",
    "df = pd.DataFrame(all_dictionaries.values()).T\n",
    "df.columns = ['{}'.format(i) for i in all_dictionaries.keys()]\n",
    "print(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nextstrain",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
